src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,252 @@
+        }
+
+        private async Task ComputeAndReportCurrentDiagnosticsAsync(
+            BufferedProgress<TReport> progress,
+            Document document,
+            CancellationToken cancellationToken)
+        {
+            // Being asked about this document for the first time.  Or being asked again and we have different
+            // diagnostics.  Compute and report the current diagnostics info for this document.
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var _ = ArrayBuilder<VSDiagnostic>.GetInstance(out var diagnostics);
+            foreach (var diagnostic in _diagnosticService.GetDiagnostics(document, includeSuppressedDiagnostics: false, cancellationToken))
+                diagnostics.Add(ConvertDiagnostic(document, text, diagnostic));
+
+            progress.Report(RecordDiagnosticReport(document, diagnostics.ToArray()));
+        }
+
+        private void HandleRemovedDocuments(DiagnosticParams[]? previousResults, BufferedProgress<TReport> progress)
+        {
+            if (previousResults == null)
+                return;
+
+            foreach (var previousResult in previousResults)
+            {
+                var textDocument = previousResult.TextDocument;
+                if (textDocument != null)
+                {
+                    var document = _solutionProvider.GetDocument(textDocument);
+                    if (document == null)
+                    {
+                        // Client is asking server about a document that no longer exists (i.e. was removed/deleted from the
+                        // workspace).  In that case we need to return an actual diagnostic report with `null` for the
+                        // diagnostics to let the client know to dump that file entirely.
+                        progress.Report(CreateReport(textDocument, diagnostics: null, resultId: null));
+                    }
+                }
+            }
+        }
+
+        private bool DiagnosticsAreUnchanged(Dictionary<Document, DiagnosticParams> documentToPreviousResult, Document document)
+        {
+            lock (_gate)
+            {
+                var workspace = document.Project.Solution.Workspace;
+                return documentToPreviousResult.TryGetValue(document, out var previousResult) &&
+                       _documentIdToLastResultId.TryGetValue((workspace, document.Id), out var lastReportedResultId) &&
+                       lastReportedResultId == previousResult.PreviousResultId;
+            }
+        }
+
+        private TReport RecordDiagnosticReport(Document document, VSDiagnostic[] diagnostics)
+        {
+            lock (_gate)
+            {
+                // Keep track of the diagnostics we reported here so that we can short-circuit producing diagnostics for
+                // the same diagnostic set in the future.
+                var resultId = _nextResultId++.ToString();
```

Comment:
this confused me for a bit until I realized that the resultId is _per document_ and not _per LSP response.  Perhaps adding a comment to nextResultId and/or renaming (e.g. _nextDocumentResultId) to make it clearer that this resultId maps to a document, and not an LSP request/response

Reply:
will do.